Concurrent threads might use uninitialised SFTP client

**In what version(s) of Spring Integration are you seeing this issue?**

spring-integration-sftp 6.0.1

**Describe the bug**

If SFTP client is not yet initialised and two threads simultaneously try to use SFTP client, one of them may fail.

`DefaultSftpSessionFactory::initClientSession` calls `DefaultSftpSessionFactory::initClient` which initialises the client.
Client initialisation is done only by the first thread.
The second thread skips initialisation and may use uninitialised/partially initialised client.
It causes exception:
```java
Caused by: java.lang.IllegalStateException: failed to create SFTP Session
	at org.springframework.integration.sftp.session.DefaultSftpSessionFactory.getSession(DefaultSftpSessionFactory.java:291)
	at org.springframework.integration.sftp.session.DefaultSftpSessionFactory.getSession(DefaultSftpSessionFactory.java:67)
	at org.springframework.integration.file.remote.session.CachingSessionFactory$1.createForPool(CachingSessionFactory.java:85)
	at org.springframework.integration.file.remote.session.CachingSessionFactory$1.createForPool(CachingSessionFactory.java:82)
	at org.springframework.integration.util.SimplePool.doGetItem(SimplePool.java:206)
	at org.springframework.integration.util.SimplePool.getItem(SimplePool.java:187)
	... 20 common frames omitted
Caused by: java.lang.IllegalStateException: SshClient not started. Please call start() method before connecting to a server
	at org.apache.sshd.client.SshClient.doConnect(SshClient.java:626)
	at org.apache.sshd.client.SshClient.doConnect(SshClient.java:615)
	at org.apache.sshd.client.SshClient.connect(SshClient.java:546)
	at org.apache.sshd.client.session.ClientSessionCreator.connect(ClientSessionCreator.java:199)
	at org.apache.sshd.client.session.ClientSessionCreator.connect(ClientSessionCreator.java:188)
	at org.springframework.integration.sftp.session.DefaultSftpSessionFactory.initClientSession(DefaultSftpSessionFactory.java:314)
	at org.springframework.integration.sftp.session.DefaultSftpSessionFactory.getSession(DefaultSftpSessionFactory.java:281)
	... 25 common frames omitted
```


**To Reproduce**

After application startup when SFTP client is not yet initialised make two SFTP requests simultaneously

**Expected behavior**

All threads must wait for client initialisation. 
Uninitialised/partially initialised client should never be used.

--- a/spring-integration-sftp/src/main/java/org/springframework/integration/sftp/session/DefaultSftpSessionFactory.java
+++ b/spring-integration-sftp/src/main/java/org/springframework/integration/sftp/session/DefaultSftpSessionFactory.java
@@ -1,5 +1,5 @@
 /*
- * Copyright 2002-2022 the original author or authors.
+ * Copyright 2002-2023 the original author or authors.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -22,7 +22,6 @@ import java.security.GeneralSecurityException;
 import java.security.KeyPair;
 import java.time.Duration;
 import java.util.Collection;
-import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.locks.Lock;
 import java.util.concurrent.locks.ReentrantLock;
 
@@ -61,6 +60,7 @@ import org.springframework.util.Assert;
  * @author David Liu
  * @author Pat Turner
  * @author Artem Bilan
+ * @author Krzysztof Debski
  *
  * @since 2.0
  */
@@ -68,7 +68,7 @@ public class DefaultSftpSessionFactory implements SessionFactory<SftpClient.DirE
 
 	private final SshClient sshClient;
 
-	private final AtomicBoolean initialized = new AtomicBoolean();
+	private volatile boolean initialized;
 
 	private final boolean isSharedSession;
 
@@ -321,41 +321,50 @@ public class DefaultSftpSessionFactory implements SessionFactory<SftpClient.DirE
 	}
 
 	private void initClient() throws IOException {
-		if (this.initialized.compareAndSet(false, true)) {
-			if (this.port <= 0) {
-				this.port = SshConstants.DEFAULT_PORT;
-			}
-			ServerKeyVerifier serverKeyVerifier =
-					this.allowUnknownKeys ? AcceptAllServerKeyVerifier.INSTANCE : RejectAllServerKeyVerifier.INSTANCE;
-			if (this.knownHosts != null) {
-				serverKeyVerifier = new ResourceKnownHostsServerKeyVerifier(this.knownHosts);
-			}
-			this.sshClient.setServerKeyVerifier(serverKeyVerifier);
-
-			this.sshClient.setPasswordIdentityProvider(PasswordIdentityProvider.wrapPasswords(this.password));
-			if (this.privateKey != null) {
-				IoResource<Resource> privateKeyResource =
-						new AbstractIoResource<>(Resource.class, this.privateKey) {
-
-							@Override
-							public InputStream openInputStream() throws IOException {
-								return getResourceValue().getInputStream();
-							}
-						};
-				try {
-					Collection<KeyPair> keys =
-							SecurityUtils.getKeyPairResourceParser()
-									.loadKeyPairs(null, privateKeyResource,
-											FilePasswordProvider.of(this.privateKeyPassphrase));
-					this.sshClient.setKeyIdentityProvider(KeyIdentityProvider.wrapKeyPairs(keys));
-				}
-				catch (GeneralSecurityException ex) {
-					throw new IOException("Cannot load private key: " + this.privateKey.getFilename(), ex);
+		if (!this.initialized) {
+			synchronized (this) {
+				if (!this.initialized) {
+					doInitClient();
+					this.initialized = true;
 				}
 			}
-			this.sshClient.setUserInteraction(this.userInteraction);
-			this.sshClient.start();
 		}
+	}
+
+	private void doInitClient() throws IOException {
+		if (this.port <= 0) {
+			this.port = SshConstants.DEFAULT_PORT;
+		}
+		ServerKeyVerifier serverKeyVerifier =
+				this.allowUnknownKeys ? AcceptAllServerKeyVerifier.INSTANCE : RejectAllServerKeyVerifier.INSTANCE;
+		if (this.knownHosts != null) {
+			serverKeyVerifier = new ResourceKnownHostsServerKeyVerifier(this.knownHosts);
+		}
+		this.sshClient.setServerKeyVerifier(serverKeyVerifier);
+
+		this.sshClient.setPasswordIdentityProvider(PasswordIdentityProvider.wrapPasswords(this.password));
+		if (this.privateKey != null) {
+			IoResource<Resource> privateKeyResource =
+					new AbstractIoResource<>(Resource.class, this.privateKey) {
+
+						@Override
+						public InputStream openInputStream() throws IOException {
+							return getResourceValue().getInputStream();
+						}
+					};
+			try {
+				Collection<KeyPair> keys =
+						SecurityUtils.getKeyPairResourceParser()
+									 .loadKeyPairs(null, privateKeyResource,
+											 FilePasswordProvider.of(this.privateKeyPassphrase));
+				this.sshClient.setKeyIdentityProvider(KeyIdentityProvider.wrapKeyPairs(keys));
+			}
+			catch (GeneralSecurityException ex) {
+				throw new IOException("Cannot load private key: " + this.privateKey.getFilename(), ex);
+			}
+		}
+		this.sshClient.setUserInteraction(this.userInteraction);
+		this.sshClient.start();
 	}
 
 	@Override

--- a/spring-integration-sftp/src/main/java/org/springframework/integration/sftp/session/DefaultSftpSessionFactory.java
+++ b/spring-integration-sftp/src/main/java/org/springframework/integration/sftp/session/DefaultSftpSessionFactory.java
@@ -355,8 +355,8 @@ public class DefaultSftpSessionFactory implements SessionFactory<SftpClient.DirE
 			try {
 				Collection<KeyPair> keys =
 						SecurityUtils.getKeyPairResourceParser()
-									 .loadKeyPairs(null, privateKeyResource,
-											 FilePasswordProvider.of(this.privateKeyPassphrase));
+									.loadKeyPairs(null, privateKeyResource,
+										FilePasswordProvider.of(this.privateKeyPassphrase));
 				this.sshClient.setKeyIdentityProvider(KeyIdentityProvider.wrapKeyPairs(keys));
 			}
 			catch (GeneralSecurityException ex) {

--- a/spring-integration-sftp/src/test/java/org/springframework/integration/sftp/session/SftpSessionFactoryTests.java
+++ b/spring-integration-sftp/src/test/java/org/springframework/integration/sftp/session/SftpSessionFactoryTests.java
@@ -1,5 +1,5 @@
 /*
- * Copyright 2014-2022 the original author or authors.
+ * Copyright 2014-2023 the original author or authors.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -17,15 +17,24 @@
 package org.springframework.integration.sftp.session;
 
 import java.io.File;
+import java.io.IOException;
 import java.net.ConnectException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
 
 import org.apache.sshd.common.SshException;
 import org.apache.sshd.server.SshServer;
 import org.apache.sshd.server.keyprovider.SimpleGeneratorHostKeyProvider;
+import org.apache.sshd.sftp.server.SftpSubsystemFactory;
 import org.junit.jupiter.api.Test;
+
+import org.springframework.core.task.AsyncTaskExecutor;
+import org.springframework.core.task.SimpleAsyncTaskExecutor;
 
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.api.Assertions.fail;
+import static org.awaitility.Awaitility.await;
 
 /**
  * @author Gary Russell
@@ -81,6 +90,39 @@ public class SftpSessionFactoryTests {
 			}
 
 			assertThat(server.getActiveSessions().size()).isEqualTo(0);
+		}
+	}
+
+	@Test
+	public void concurrentGetSessionDoesntCauseFailure() throws IOException {
+		try (SshServer server = SshServer.setUpDefaultServer()) {
+			server.setPasswordAuthenticator((arg0, arg1, arg2) -> true);
+			server.setPort(0);
+			server.setKeyPairProvider(new SimpleGeneratorHostKeyProvider(new File("hostkey.ser").toPath()));
+			server.setSubsystemFactories(Collections.singletonList(new SftpSubsystemFactory()));
+			server.start();
+
+			DefaultSftpSessionFactory sftpSessionFactory = new DefaultSftpSessionFactory();
+			sftpSessionFactory.setHost("localhost");
+			sftpSessionFactory.setPort(server.getPort());
+			sftpSessionFactory.setUser("user");
+			sftpSessionFactory.setPassword("pass");
+			sftpSessionFactory.setAllowUnknownKeys(true);
+
+			List<SftpSession> concurrentSessions = new ArrayList<>();
+
+			AsyncTaskExecutor asyncTaskExecutor = new SimpleAsyncTaskExecutor();
+			for (int i = 0; i < 3; i++) {
+				asyncTaskExecutor.execute(() -> concurrentSessions.add(sftpSessionFactory.getSession()));
+			}
+
+			await().until(() -> concurrentSessions.size() == 3);
+
+			assertThat(concurrentSessions.get(0))
+					.isNotEqualTo(concurrentSessions.get(1))
+					.isNotEqualTo(concurrentSessions.get(2));
+
+			assertThat(concurrentSessions.get(1)).isNotEqualTo(concurrentSessions.get(2));
 		}
 	}
 

